crates/eips/src/eip1898.rs

```diff
@@ -0,0 +1,471 @@
+        matches!(self, BlockNumberOrTag::Safe)
+    }
+
+    /// Returns `true` if it's "pending"
+    pub const fn is_pending(&self) -> bool {
+        matches!(self, BlockNumberOrTag::Pending)
+    }
+
+    /// Returns `true` if it's "earliest"
+    pub const fn is_earliest(&self) -> bool {
+        matches!(self, BlockNumberOrTag::Earliest)
+    }
+}
+
+impl From<u64> for BlockNumberOrTag {
+    fn from(num: u64) -> Self {
+        BlockNumberOrTag::Number(num)
+    }
+}
+
+impl From<U64> for BlockNumberOrTag {
+    fn from(num: U64) -> Self {
+        num.to::<u64>().into()
+    }
+}
+
+impl Serialize for BlockNumberOrTag {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match *self {
+            BlockNumberOrTag::Number(x) => serializer.serialize_str(&format!("0x{x:x}")),
+            BlockNumberOrTag::Latest => serializer.serialize_str("latest"),
+            BlockNumberOrTag::Finalized => serializer.serialize_str("finalized"),
+            BlockNumberOrTag::Safe => serializer.serialize_str("safe"),
+            BlockNumberOrTag::Earliest => serializer.serialize_str("earliest"),
+            BlockNumberOrTag::Pending => serializer.serialize_str("pending"),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for BlockNumberOrTag {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?.to_lowercase();
+        s.parse().map_err(serde::de::Error::custom)
+    }
+}
+
+impl FromStr for BlockNumberOrTag {
+    type Err = ParseBlockNumberError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let block = match s {
+            "latest" => Self::Latest,
+            "finalized" => Self::Finalized,
+            "safe" => Self::Safe,
+            "earliest" => Self::Earliest,
+            "pending" => Self::Pending,
+            _number => {
+                if let Some(hex_val) = s.strip_prefix("0x") {
+                    let number = u64::from_str_radix(hex_val, 16);
+                    BlockNumberOrTag::Number(number?)
+                } else {
+                    return Err(HexStringMissingPrefixError::default().into());
+                }
+            }
+        };
+        Ok(block)
+    }
+}
+
+impl fmt::Display for BlockNumberOrTag {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            BlockNumberOrTag::Number(x) => write!(f, "0x{x:x}"),
+            BlockNumberOrTag::Latest => f.write_str("latest"),
+            BlockNumberOrTag::Finalized => f.write_str("finalized"),
+            BlockNumberOrTag::Safe => f.write_str("safe"),
+            BlockNumberOrTag::Earliest => f.write_str("earliest"),
+            BlockNumberOrTag::Pending => f.write_str("pending"),
+        }
+    }
+}
+
+/// Error variants when parsing a [BlockNumberOrTag]
+#[derive(Debug, thiserror::Error)]
```

Comment:
this now needs to manually impl Display + all the From impls

and error trait must be implemented manually with a cfg(feature = "std)